sys/mips/rootdev.c

```diff
@@ -2,73 +2,168 @@
-}
+static resource_t *rootdev_alloc_resource(device_t *dev, res_type_t type,
+                                          int rid, rman_addr_t start,
+                                          rman_addr_t end, size_t size,
+                                          res_flags_t flags) {
+  rootdev_t *rd = dev->parent->state;
+  rman_t *rman = NULL;
+  resource_t *r;
 
-static resource_t *rootdev_alloc_resource(device_t *bus, device_t *child,
-                                          res_type_t type, int rid,
-                                          rman_addr_t start, rman_addr_t end,
-                                          size_t size, res_flags_t flags) {
+  if (type == RT_MEMORY)
+    rman = &rd->mem;
+  else if (type == RT_IRQ)
+    rman = &rd->irq;
+  else
+    panic("Resource type not handled!");
 
-  resource_t *r =
-    rman_alloc_resource(&rm_mem, start, end, size, 1, RF_NONE, child);
+  if (!(r = rman_alloc_resource(rman, start, end, size, 1, RF_NONE)))
+    return NULL;
 
-  if (r) {
+  if (type == RT_MEMORY) {
     r->r_bus_tag = generic_bus_space;
-    if (flags & RF_ACTIVE)
-      bus_space_map(r->r_bus_tag, r->r_start, r->r_end - r->r_start + 1,
-                    &r->r_bus_handle);
+    r->r_bus_handle = r->r_start;
+
+    if (flags & RF_ACTIVE) {
+      if (bus_activate_resource(dev, type, rid, r)) {
+        rman_release_resource(r);
+        return NULL;
+      }
+    }
   }
 
   return r;
 }
 
+static void rootdev_release_resource(device_t *dev, res_type_t type, int rid,
+                                     resource_t *r) {
+  panic("not implemented!");
+}
+
+static int rootdev_activate_resource(device_t *dev, res_type_t type, int rid,
+                                     resource_t *r) {
+  int error = 0;
+
+  if (r->r_flags & RF_ACTIVE)
+    return 0;
+
+  if (type == RT_MEMORY)
+    error = bus_space_map(r->r_bus_tag, r->r_bus_handle, rman_get_size(r),
+                          &r->r_bus_handle);
+
+  if (error == 0)
+    rman_activate_resource(r);
+
+  return error;
+}
+
+static void rootdev_intr_handler(ctx_t *ctx, device_t *dev, void *arg) {
+  rootdev_t *rd = dev->state;
+  unsigned pending = (_REG(ctx, CAUSE) & _REG(ctx, SR)) & CR_IP_MASK;
+
+  for (int i = 7; i >= 0; i--) {
+    unsigned irq = CR_IP0 << i;
+
+    if (pending & irq) {
+      intr_event_run_handlers(&rd->intr_event[i]);
+      pending &= ~irq;
+    }
+  }
+}
+
+static int rootdev_probe(device_t *bus) {
+  return 1;
+}
+
+static int rootdev_attach(device_t *bus) {
+  rootdev_t *rd = bus->state;
+
+  /* Manages space occupied by I/O devices: PCI, FPGA, system controler, ... */
+  rman_init(&rd->mem, "Malta I/O space", 0x10000000, 0x1fffffff, RT_MEMORY);
+  rman_init(&rd->irq, "MIPS interrupts", 0, 7, RT_IRQ);
+
+#define MIPS_INTR_EVENT(irq, name)                                             \
+  intr_event_init(&rd->intr_event[irq], irq, name, rootdev_mask_irq,           \
+                  rootdev_unmask_irq, NULL)
```

Comment:
I prefer to change it into `MIPS_INTR_EVENT(rd, irq, name)` or even create a separate function for that purpose.

Reply:
Fixed.